intercom-common/src/model/cominterface.rs

```diff
@@ -0,0 +1,347 @@
+        //
+        // In practice the NO_BASE should be used ONLY for the IUnknown itself.
+        let base = attr.base()
+                .map_err( |msg| ParseError::ComInterface(
+                    item.get_ident().unwrap().to_string(), msg ) )?;
+        let base = match base {
+            Some( b ) => if b == "NO_BASE" { None } else { Some( b.to_owned() ) },
+            None => Some( Ident::new( "IUnknown", Span::call_site() ) ),
+        };
+
+        // Visibility for trait interfaces is the visibility of the trait.
+        //
+        // For implicit interfaces (impl Struct) the visibility is always public.
+        // These interfaces should only exist for COM types that are meant to be
+        // called from external sources as they can't be impl'd for random ComItf.
+        //
+        // Note this may conflict with visibility of the actual [com_class], but
+        // nothing we can do for this really.
+        let visibility = if let ::syn::Item::Trait( ref t ) = *item {
+                    t.vis.clone()
+                } else {
+                    parse_quote!( pub )
+                };
+
+        let variants = OrderMap::from_iter(
+            [ TypeSystem::Automation, TypeSystem::Raw ].into_iter().map( |&ts| {
+
+            let itf_unique_ident = Ident::new( 
+                    &format!( "{}_{:?}", itf_ident.to_string(), ts ), Span::call_site() );
+
+                // IUnknown interfaces do not have type system variants.
+                let unique_base = match base {
+                    Some( ref iunk ) if iunk == "IUnknown" => base.clone(),
+                    ref b => b.as_ref().map( |b| Ident::new( &format!( "{}_{:?}", b, ts ), Span::call_site() ) )
+                };
+
+                let iid_attr = attr.iid( ts )
+                        .map_err( |msg| ParseError::ComInterface(
+                            item.get_ident().unwrap().to_string(), msg ) )?;
+                let iid = match iid_attr {
+                    Some( iid ) => GUID::parse( &iid.value() )
+                        .map_err( |_| ParseError::ComInterface(
+                                item.get_ident().unwrap().to_string(),
+                                "Bad IID format".into() ) )?,
+                    None => ::utils::generate_iid(
+                            crate_name, &itf_unique_ident.to_string(), ts )
+                };
+
+                // Read the method details.
+                //
+                // TODO: Currently we ignore invalid methods. We should probably do
+                //       something smarter.
+                let methods = fns.iter()
+                        .map( | sig |
+                            ComMethodInfo::new( sig, ts ) )
+                        .filter_map( |r| r.ok() )
+                        .collect::<Vec<_>>();
+
+                Ok( ( ts, ComInterfaceVariant {
+                    display_name: itf_ident.clone(),
+                    unique_name : itf_unique_ident,
```

Comment:
IID-declarations being often global  on Windows, should we provide an option to include the name of the library in all types?